### ISSUE TITLE
Directive resolver isAuthenticated does not stop the request if the user is not authenticated [SOLVED]

### ISSUE THREAD

**Issue by leonelpena** [LawJolla/prisma-auth0-example#8]
Hello,

First of all thank you for your article and code example both has been very useful for me. 

As I write in the title, the directive resolver does not stop the request when there is no user (no header authenticated) and it continues to the resolver, I think I am missing something...

Schema:

```
directive @isAuthenticated on QUERY | FIELD | MUTATION
directive @hasRole(roles: [String]) on QUERY | FIELD | MUTATION

type Query {
	user(id: Int): User @hasRole(roles: ["ADMIN"])
	allUsers: [User] @hasRole(roles: ["ADMIN"])
	currentUser: User @isAuthenticated
}
```

Directive resolver:
```
const isLoggedIn = ctx => {
  const user = ctxUser(ctx, userLocationOnContext);
  if (!user) {
    throw new NotLoggedInError();
  }
  return user;
}

const directiveResolvers = {
  isAuthenticated(result, source, args, ctx) {
    isLoggedIn(ctx);
    return result;
  },
}
```

The function **ctxUser** is the same as yours and the class **NotLoggedInError** is implemented as follows:
```
import { createError } from "apollo-errors";

export const NotLoggedInError = createError('NotLoggedInError', {
  message: 'Not logged in.'
});
```

The user resolver is:

```
currentUser(root, args, context) {
   return User.find({ where: context.req.user.id });
},
```

The getUser middleware is:

```
const getUser = async (req, res, next) => {
  const token = getToken(_get(req, "headers.authorization"));
  if (token) {
    try {
      const payload = validateAndParseTokenPayload(token);
      const user = await User.findById(payload.id);
      req.user = { id: payload.id, role: user.role, token: req.headers.authorization };
    } catch (err) {
      res.statusCode = 401;
      return res.end('{"errors": [{"message": "token damaged"}]}');
    }
  }
  next();
}
```

If I execute the query:
```
query {
  currentUser {
    id,
    firstName,
    lastName,
    email,
    role
  }
}
```

I get the output:
```
TypeError: Cannot read property 'id' of undefined
    at Object.currentUser (/Users/usuario/Music/mercado-capacitacion/graphql-server/src/resolvers/userResolvers.js:15:50)
```

It seems like the directive resolver does not stop and the user resolvers is executed normally.

**Comment by LawJolla** [comment]
Hi @leonelpena ,

Thanks for your kind words and question!

Who things.  First, there's a problem in your directive resolver.  The arguments are:  `next, source, args, ctx`.  To pass the resolver (like middleware), you need to invoke the next function.

```
isAuthenticated(result, source, args, ctx) {
    isLoggedIn(ctx);
    return result;
  }
```

Becomes
```
isAuthenticated(result, source, args, ctx) {
    isLoggedIn(ctx);
    return result();
  }
```
By naming it `result`, it makes sense to return it as an object.  But it's actually a `next` style function.

Second, I haven't used `apollo-errors` yet, but it doesn't look like throwing it works as intended.  A JS error thrown in a resolver stops execution of that resolver and child resolvers.  Try `isLoggedIn` with a regular error and see if that changes things.

```
const isLoggedIn = ctx => {
  const user = ctxUser(ctx, userLocationOnContext);
  if (!user) throw new Error('Not logged in');
  return user;
}
```

**Comment by leonelpena** [comment]
If I change the result argument for next(), I get this error: 
`"message": "next is not a function"`
and if I changed the apollo-error for a normal error the execution does not stop in the directive resolver.

I added a ```console.log(`user: ${user}`)``` on the isLoggedIn function and I realized that it does not appeared on the console, is it possible that the isAuthenticated directive resolver is not been executed?

This is the function to attach directives:
```
const attachDirectives = schema => {
  forEachField(schema, field => {
    const directives = field.astNode.directives;
    directives.forEach(directive => {
      const directiveName = directive.name.value;
      const resolver = directiveResolvers[directiveName];

      if (resolver) {
        const oldResolve = field.resolve;
        const Directive = schema.getDirective(directiveName);
        const args = getArgumentValues(Directive, directive);

        field.resolve = function() {
          const [source, _, context, info] = arguments;
          let promise = oldResolve.call(field, ...arguments);

          const isPrimitive = !(promise instanceof Promise);
          if (isPrimitive) {
            promise = Promise.resolve(promise);
          }

          return promise.then(result =>
            resolver(result, source, args, context, info)
          );
        };
      }
    });
  });
};
```

**Comment by LawJolla** [comment]
I'm not sure then.  You're using packages with which I'm unfamilar and have substantially modified the original repo.  `user` is clearly not falsy for whatever reason.

My suggestion would be to unwind your server to match the demo, get it working, and then rebuild your features.  Sorry that I can't be of more help.

**Comment by leonelpena** [comment]
Ok, I will tried it. Thank you for your help.

**Comment by leonelpena** [comment]
I have solved it. The problem was the function that I used to attach directive resolvers (`attachDirectives`), which I got from another tutorial, was no working as expected, at least not the way I was using it. This was the fix:
1. I commented the function and included the directive resolvers directly into the `makeExecutableSchema` function, just like your example.
1. In the directive resolvers I change the `result` argument to `next` and used as you said: `next()`.

That was all. @LawJolla thank you for your help!
